File: Semester 4/Class12_Computer_Contest_Level_3_Notes_and_Homework_Nov_26/lvl3_hw_sol/death_gun.java
import java.util.*;
import java.io.*;
public class homework {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    public static void main(String[] args) throws IOException{
    	int n = readInt(), idx = 0;
    	List<Integer> adj[] = new ArrayList[2*n];
    	for(int i=0; i<adj.length; i++)
    		adj[i] = new ArrayList();
    	int [] indeg = new int[2*n];
    	Map<String, Integer> map = new HashMap();
    	List<String> name = new ArrayList();
    	for(int i=1; i<=n; i++) {
    		String a = next(), b = next();
    		if(!map.containsKey(a)) {
    			map.put(a, idx); name.add(a); idx++;
    		}
    		if(!map.containsKey(b)) {
    			map.put(b, idx); name.add(b); idx++;
    		}
    		adj[map.get(b)].add(map.get(a)); 
    		indeg[map.get(a)]++;
    	}
    	PriorityQueue<Integer> q = new PriorityQueue<>();
    	List<Integer> order = new ArrayList();
    	for(int i=0; i<idx; i++) {
    		if(indeg[i] == 0) q.add(i);
     	}
    	while(!q.isEmpty()) {
    		int cur = q.poll();  order.add(cur);
    		for(int nxt: adj[cur]) {
    			indeg[nxt]--;
    			if(indeg[nxt] == 0) q.add(nxt);
    		}
    	}
    	for(int x : order) System.out.println(name.get(x));
    }
    static String next () throws IOException {
        while (st == null || ! st.hasMoreTokens())
            st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }
    static long readLong () throws IOException {
        return Long.parseLong(next());
    }
    static int readInt () throws IOException {
        return Integer.parseInt(next());
    }
    static double readDouble () throws IOException {
        return Double.parseDouble(next());
    }
    static char readCharacter () throws IOException {
        return next().charAt(0);
    }  
    static String readLine () throws IOException {
        return br.readLine().trim();
    }
}

File: Semester 4/Class12_Computer_Contest_Level_3_Notes_and_Homework_Nov_26/lvl3_hw_sol/death_gun.py
import sys
import heapq
input = sys.stdin.readline
n = int(input())
adj = [[] for i in range(2*n)]
d, name, indeg = {}, [], [0]*(2*n)
for i in range(n):
    a, b = input().split()
    if a not in d:
        d[a] = len(name)
        name.append(a)
    if b not in d:
        d[b] = len(name)
        name.append(b)
    adj[d[b]].append(d[a])
    indeg[d[a]] += 1
q, ans = [], []
for i in range(len(name)):
    if indeg[i] == 0:
        q.append(i)
heapq.heapify(q)
while q:
    cur = heapq.heappop(q)
    ans.append(cur)
    for nxt in adj[cur]:
        indeg[nxt] -= 1
        if indeg[nxt] == 0:
            heapq.heappush(q, nxt)
for i in ans:
    print(name[i])